### ISSUE TITLE
[BUG] inconsistent results for fix addforce during minimization when using kokkos/cuda

### ISSUE THREAD

**Issue by nrbertin** [lammps/lammps#3759]
**Summary**

Using `fix addforce` during `minimize` seems not to be working when using kokkos/cuda (`-k on g 1 -sf kk -pk kokkos`). Everything works as expected when using the same lammps binary with kokkos off.

**LAMMPS Version and Platform**

LAMMPS (8 Feb 2023) compiled with kokkos (Cuda,OpenMP)
gcc/8.3.1
cuda/11.2.0
IBM Power9 (lassen)

**Expected Behavior**

Here is a minimal input script that shows the issue. A perfect crystal with free surfaces along the z direction is created, then a stress of 100 MPa is applied to the free surface atoms using the `fix addforce` and the energy is minimized under this force:

```
units metal
atom_style atomic
boundary p p s

variable a0 equal 3.30263301805426
lattice bcc ${a0} orient x 1 1 1 orient y -1 2 -1 orient z -1 0 1
region box block 0 3 0 18 0 24 units lattice
create_box 1 box
create_atoms 1 box

change_box all triclinic

pair_style   	eam/alloy
pair_coeff   	* * newPP1_47-setfl.eam.alloy Ta
mass         	* 180.9479
neighbor        0.7 bin
neigh_modify    every 1 delay 0

# Identify surface atoms
variable boundary_thickness equal 4.0
variable zlo0 equal "zlo+v_boundary_thickness"
variable zhi0 equal "zhi-v_boundary_thickness"
region upper block INF INF INF INF ${zhi0} INF units box
region lower block INF INF INF INF INF ${zlo0} units box
group upper region upper
group lower region lower
group surfaces union upper lower

# Identify central (bulk) atoms awy from the free surfaces
variable boundary_thickness equal 16.0
variable zlo0 equal "zlo+v_boundary_thickness"
variable zhi0 equal "zhi-v_boundary_thickness"
region bulk block INF INF INF INF ${zlo0} ${zhi0} units box
group bulk region bulk

compute 1 bulk pe/atom
compute ebulk bulk reduce sum c_1

compute 2 bulk stress/atom NULL virial
compute pbulk bulk reduce ave c_2[*]
variable Sxzmpa equal "0.1/0.5/v_a0^3*c_pbulk[5]"

thermo 100
thermo_style custom step pe c_ebulk v_Sxzmpa

# Apply shear tau (MPa)
variable tau equal 100.0
variable A equal "lx*ly"
variable energyConv equal 160217.662089
variable fx_upper equal "v_A/v_energyConv/count(upper)*v_tau"
variable fx_lower equal "-v_A/v_energyConv/count(lower)*v_tau"

fix 1 upper addforce ${fx_upper} 0.0 0.0
fix 2 lower addforce ${fx_lower} 0.0 0.0
fix_modify 1 energy yes
fix_modify 2 energy yes

minimize 0.0 1.0e-8 10000 10000
```

When running the above script with kokkos off
`./lmp_8Feb23_lassen_kokkos -in shear.in`
the correct following output is obtained:

```
LAMMPS (8 Feb 2023)
OMP_NUM_THREADS environment is not set. Defaulting to 1 thread. (../comm.cpp:98)
  using 1 OpenMP thread(s) per MPI task
Lattice spacing in x,y,z = 5.7203282 5.3931771 4.6706284
Created orthogonal box = (0 0 0) to (17.160985 97.077188 112.09508)
  1 by 1 by 1 MPI processor grid
Created 10584 atoms
  using lattice units in orthogonal box = (0 0 -0.011209508) to (17.160985 97.077188 112.10629)
  create_atoms CPU = 0.005 seconds
Changing box ...
  triclinic box = (0 0 -0.022421258) to (17.160985 97.077188 112.1175) with tilt (0 0 0)
432 atoms in group upper
432 atoms in group lower
864 atoms in group surfaces
7560 atoms in group bulk
Neighbor list info ...
  update: every = 1 steps, delay = 0 steps, check = yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 4.6837119
  ghost atom cutoff = 4.6837119
  binsize = 2.3418559, bins = 8 42 48
  1 neighbor lists, perpetual/occasional/extra = 1 0 0
  (1) pair eam/alloy, perpetual
      attributes: half, newton on
      pair build: half/bin/newton/tri
      stencil: half/bin/3d/tri
      bin: standard
Setting up cg style minimization ...
  Unit style    : metal
  Current step  : 0
Per MPI rank memory allocation (min/avg/max) = 10.83 | 10.83 | 10.83 Mbytes
   Step         PotEng        c_ebulk        v_Sxzmpa   
         0  -85247.896     -61152.841     -3.7522879e-14
       100  -85248.675     -61152.77       100.23735    
       125  -85248.675     -61152.77       99.990348    
Loop time of 1.38537 on 1 procs for 125 steps with 10584 atoms

100.0% CPU use with 1 MPI tasks x 1 OpenMP threads

Minimization stats:
  Stopping criterion = linesearch alpha is zero
  Energy initial, next-to-last, final = 
     -85247.8957987199  -85248.6752707938  -85248.6752707938
  Force two-norm initial, final = 4.0820494 8.623314e-05
  Force max component initial, final = 0.13885328 1.8808693e-06
  Final line search alpha, max atom move = 0.5 9.4043466e-07
  Iterations, force evaluations = 125 251

MPI task timing breakdown:
Section |  min time  |  avg time  |  max time  |%varavg| %total
---------------------------------------------------------------
Pair    | 1.3162     | 1.3162     | 1.3162     |   0.0 | 95.01
Neigh   | 0          | 0          | 0          |   0.0 |  0.00
Comm    | 0.0090277  | 0.0090277  | 0.0090277  |   0.0 |  0.65
Output  | 0.00052932 | 0.00052932 | 0.00052932 |   0.0 |  0.04
Modify  | 0.0074873  | 0.0074873  | 0.0074873  |   0.0 |  0.54
Other   |            | 0.05212    |            |       |  3.76

Nlocal:          10584 ave       10584 max       10584 min
Histogram: 1 0 0 0 0 0 0 0 0 0
Nghost:           6831 ave        6831 max        6831 min
Histogram: 1 0 0 0 0 0 0 0 0 0
Neighs:         135432 ave      135432 max      135432 min
Histogram: 1 0 0 0 0 0 0 0 0 0

Total # of neighbors = 135432
Ave neighs/atom = 12.795918
Neighbor list builds = 0
Dangerous builds = 0
Total wall time: 0:00:01
```

As expected, the shear stress Sxzmpa ~ 100 MPa after minimization (125 steps).

**Actual Behavior**

When running the same script with kokkos on:
`./lmp_8Feb23_lassen_kokkos -k on g 1 -sf kk -pk kokkos -in shear.in`
a different output is obtained:

```
LAMMPS (8 Feb 2023)
KOKKOS mode is enabled (../kokkos.cpp:106)
  will use up to 1 GPU(s) per node
WARNING: When using a single thread, the Kokkos Serial backend (i.e. Makefile.kokkos_mpi_only) gives better performance than the OpenMP backend (../kokkos.cpp:227)
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
  using 1 OpenMP thread(s) per MPI task
Lattice spacing in x,y,z = 5.7203282 5.3931771 4.6706284
Created orthogonal box = (0 0 0) to (17.160985 97.077188 112.09508)
  1 by 1 by 1 MPI processor grid
Created 10584 atoms
  using lattice units in orthogonal box = (0 0 -0.011209508) to (17.160985 97.077188 112.10629)
  create_atoms CPU = 0.012 seconds
Changing box ...
  triclinic box = (0 0 -0.022421258) to (17.160985 97.077188 112.1175) with tilt (0 0 0)
432 atoms in group upper
432 atoms in group lower
864 atoms in group surfaces
7560 atoms in group bulk
Neighbor list info ...
  update: every = 1 steps, delay = 0 steps, check = yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 4.6837119
  ghost atom cutoff = 4.6837119
  binsize = 4.6837119, bins = 4 21 24
  1 neighbor lists, perpetual/occasional/extra = 1 0 0
  (1) pair eam/alloy/kk, perpetual
      attributes: full, newton off, kokkos_device
      pair build: full/bin/kk/device
      stencil: full/bin/3d
      bin: kk/device
Setting up cg/kk style minimization ...
  Unit style    : metal
  Current step  : 0
WARNING: Fixes cannot yet send exchange data in Kokkos communication, switching to classic exchange/border communication (../comm_kokkos.cpp:651)
Per MPI rank memory allocation (min/avg/max) = 6.131 | 6.131 | 6.131 Mbytes
   Step         PotEng        c_ebulk        v_Sxzmpa   
         0  -85247.896     -61152.841     -2.8713319e-14
         8  -85248.577     -61152.839      0.00066831465
Loop time of 0.0770231 on 1 procs for 8 steps with 10584 atoms

75.9% CPU use with 1 MPI tasks x 1 OpenMP threads

Minimization stats:
  Stopping criterion = linesearch alpha is zero
  Energy initial, next-to-last, final = 
     -85247.8957989135  -85248.5774617852  -85248.5774617852
  Force two-norm initial, final = 4.0820494 0.13713473
  Force max component initial, final = 0.13885328 0.0036317902
  Final line search alpha, max atom move = 4.7683716e-07 1.7317725e-09
  Iterations, force evaluations = 8 57

MPI task timing breakdown:
Section |  min time  |  avg time  |  max time  |%varavg| %total
---------------------------------------------------------------
Pair    | 0.04309    | 0.04309    | 0.04309    |   0.0 | 55.94
Neigh   | 0          | 0          | 0          |   0.0 |  0.00
Comm    | 0.0015846  | 0.0015846  | 0.0015846  |   0.0 |  2.06
Output  | 0          | 0          | 0          |   0.0 |  0.00
Modify  | 0.017031   | 0.017031   | 0.017031   |   0.0 | 22.11
Other   |            | 0.01532    |            |       | 19.89

Nlocal:          10584 ave       10584 max       10584 min
Histogram: 1 0 0 0 0 0 0 0 0 0
Nghost:           6831 ave        6831 max        6831 min
Histogram: 1 0 0 0 0 0 0 0 0 0
Neighs:              0 ave           0 max           0 min
Histogram: 1 0 0 0 0 0 0 0 0 0
FullNghs:       270864 ave      270864 max      270864 min
Histogram: 1 0 0 0 0 0 0 0 0 0

Total # of neighbors = 270864
Ave neighs/atom = 25.591837
Neighbor list builds = 0
Dangerous builds = 0
Total wall time: 0:00:00
```

Here the minimization only proceeds for 8 steps and the final stress is still Sxzmpa ~ 0 MPa, as if `fix addforce` was not used.

In summary, running the same script with kokkos on or off produces different results, with kokkos/off results being as expected and kokkos/on results being as if the `fix addforce` was ignored.

**Steps to Reproduce**

Using the above input script with the EAM potential file whose link is provided below.

**Further Information, Files, and Links**

Interatomic potential:
https://www.ctcms.nist.gov/potentials/Download/2003--Li-Y-Siegel-D-J-Adams-J-B-Liu-X-Y--Ta/2/newPP1_47-setfl.eam.alloy

**Comment by stanmoore1** [comment]
@nrbertin thanks for the bug report, looking at this now.

**Comment by stanmoore1** [comment]
I can reproduce the issue. It runs fine with UVM so the issue is a missing sync/modify for data movement.

**Comment by stanmoore1** [comment]
@nrbertin this should be fixed with #3766. I still get a different number of minimize iterations with Kokkos vs non-Kokkos but as far as I can tell it is related to accumulation of roundoff error. Let us know if you have more issues.

**Comment by stanmoore1** [comment]
If I randomly displace the atoms slightly (e.g. 1e-8) then I get more similar behavior between Kokkos and non-Kokkos.

```
displace_atoms all random 1e-8 1e-8 1e-8 123456
```

**Comment by nrbertin** [comment]
I tested it from [stanmoore1:kk_min_bug](https://github.com/stanmoore1/lammps/tree/kk_min_bug) and everything works fine now. Thanks for the quick bugfix!
